### PR TITLE
Add "allowReconnect" to SignalR CloseMessages

### PR DESCRIPTION
**Pull request by halter73**
Add "allowReconnect" to SignalR CloseMessages so the Azure SignalR Service can disconnect clients with a reason while still allowing the automatic reconnect features of the C# and TS/JS client to automatically reconnect.

For backwards compatibility, received CloseMessages with "allowReconnect" completely missing is interpreted the same as if it were false.

Compat Testing:
* [x] Tested 3.0 JS Client receiving close message with this new flag from 3.1 Server with JSON protocol
* [x] Tested 3.0 JS Client receiving close message with this new flag from 3.1 Server with MessagePack protocol
* [x] Tested 3.0 .NET Client receiving close message with this new flag from 3.1 Server with JSON protocol
* [x] Tested 3.0 .NET Client receiving close message with this new flag from 3.1 Server with MessagePack protocol
* [x] Tested 3.1 JS Client receiving close message *without* this new flag from 3.0 Server with MessagePack protocol
* [x] Tested 3.1 .NET Client receiving close message *without* this new flag from 3.0 Server with MessagePack protocol

### CODE REVIEW

File: src/SignalR/server/Core/src/HubConnectionContext.cs
Comment (halter73) [inline src/SignalR/server/Core/src/HubConnectionContext.cs]: This to me is the big remaining question.

File: src/SignalR/common/SignalR.Common/ref/Microsoft.AspNetCore.SignalR.Common.netcoreapp.cs
Comment (halter73) [inline src/SignalR/common/SignalR.Common/ref/Microsoft.AspNetCore.SignalR.Common.netcoreapp.cs]: Note the new API @davidfowl

**Comment by anurse** [comment]
As a reminder: Do not merge until I post back indicating we have approval from the release QB

File: src/SignalR/common/Protocols.MessagePack/src/Protocol/MessagePackHubProtocol.cs
Comment (BrennanConroy) [inline src/SignalR/common/Protocols.MessagePack/src/Protocol/MessagePackHubProtocol.cs]: default should be false?

File: src/SignalR/common/Protocols.MessagePack/src/Protocol/MessagePackHubProtocol.cs
Comment (BrennanConroy) [inline src/SignalR/common/Protocols.MessagePack/src/Protocol/MessagePackHubProtocol.cs]: super nit: we could omit this if it's the default value (false) and save on message size

File: src/SignalR/common/Shared/JsonUtils.cs
Comment (BrennanConroy) [inline src/SignalR/common/Shared/JsonUtils.cs]: I believe that Newtonsoft.Json is too nice and will read both `"true"` and `true` as boolean, but we'd prefer to be stricter and only allow `true`

File: src/SignalR/common/Shared/SystemTextJsonExtensions.cs
Comment (BrennanConroy) [inline src/SignalR/common/Shared/SystemTextJsonExtensions.cs]: woah, is this that new pattern matching feature

+Pranav Points

File: src/SignalR/clients/ts/signalr-protocol-msgpack/src/MessagePackHubProtocol.ts
Comment (BrennanConroy) [inline src/SignalR/clients/ts/signalr-protocol-msgpack/src/MessagePackHubProtocol.ts]: We should normalize this name, both "allowReconnect" and "allowAutomaticReconnect" are used throughout this PR

File: src/SignalR/server/Core/src/HubConnectionContext.cs
Comment (BrennanConroy) [inline src/SignalR/server/Core/src/HubConnectionContext.cs]: nit: this name isn't great, it doesn't really do anything specific to allowing reconnect

File: src/SignalR/server/Core/src/HubConnectionHandler.cs
Comment (BrennanConroy) [inline src/SignalR/server/Core/src/HubConnectionHandler.cs]: nit:
```suggestion
                closeMessage = new CloseMessage(error: null, allowAutomaticReconnect);
```

**Review by BrennanConroy — COMMENTED** [review]
We should also update the spec doc

Tests should be added to https://github.com/aspnet/AspNetCore/blob/9557630c0a470618b30679e8effcf019b1c74164/src/SignalR/common/SignalR.Common/test/Internal/Protocol/JsonHubProtocolTestsBase.cs#L36 and https://github.com/aspnet/AspNetCore/blob/9557630c0a470618b30679e8effcf019b1c74164/src/SignalR/common/SignalR.Common/test/Internal/Protocol/MessagePackHubProtocolTestBase.cs#L63

File: src/SignalR/server/Core/src/HubConnectionContext.cs
Comment (halter73) [reply src/SignalR/server/Core/src/HubConnectionContext.cs]: What would you call this?

Comment (BrennanConroy) [reply src/SignalR/server/Core/src/HubConnectionContext.cs]: AbortInner?

File: src/SignalR/clients/ts/signalr/tests/HubConnection.test.ts
Comment (halter73) [reply src/SignalR/clients/ts/signalr/tests/HubConnection.test.ts]: This is guaranteed the way way TestConnection is implemented, since closing the TestConnection invokes the closed hander inline.

File: src/SignalR/server/Core/src/HubConnectionContext.cs
Comment (halter73) [reply src/SignalR/server/Core/src/HubConnectionContext.cs]: That feels lazy to me. Might as well call it AbortCore(). At least this name indicates how it's different from a normal Abort().

**Comment by anurse** [comment]
Put "NO MERGE" label on just to make clear that we're pending approval for 3.1.

**Comment by anurse** [comment]
Approved for 3.1. Please explcitly confirm that cross-version compat testing of this functionality has been done (3.0 client to 3.1 server, 3.1 client to 3.0 server). I've added a checklist you can use to do that to the main description.

Approval confirmation: https://devdiv.visualstudio.com/DevDiv/_workitems/edit/1001918/

File: src/SignalR/clients/ts/signalr/src/HubConnection.ts
Comment (BrennanConroy) [inline src/SignalR/clients/ts/signalr/src/HubConnection.ts]: How does this interact with someone calling stop on the hub connection?

**Review by BrennanConroy — APPROVED** [review]
I'd like the protocol spec to be updated too.

File: src/SignalR/clients/ts/signalr/src/HubConnection.ts
Comment (halter73) [reply src/SignalR/clients/ts/signalr/src/HubConnection.ts]: HubConnection.stop() immediately moves the typescript client into the disconnecting state. If HubConnection.stop() were to be called before the HubConnection's connectionClosed handler fired, the connectionClosed handler would see that the HubConnection already transitioned to the disconnecting state and transition to the disconnected state instead of trying to reconnect.

https://github.com/aspnet/AspNetCore/blob/1599a5a2a791def8a23784756c303a914c0e1631/src/SignalR/clients/ts/signalr/src/HubConnection.ts#L660

This already worked because the client needs to be able to reliably handle a connection that just happens to disconnect for any reason (not just close messages) right as someone calls HubConnection.stop()

File: src/SignalR/common/Protocols.MessagePack/src/Protocol/MessagePackHubProtocol.cs
Comment (halter73) [reply src/SignalR/common/Protocols.MessagePack/src/Protocol/MessagePackHubProtocol.cs]: Since this is message pack, and it only adds a byte to a message that should be sent at most once per connection, I'm leaving it as is for simplicity.

File: src/SignalR/common/SignalR.Common/ref/Microsoft.AspNetCore.SignalR.Common.netcoreapp.cs
Comment (halter73) [inline src/SignalR/common/SignalR.Common/ref/Microsoft.AspNetCore.SignalR.Common.netcoreapp.cs]: @aspnet/asp-net-api-reviews @davidfowl What do I need to do to get the api-approved label on this PR? I want to merge this soon since it for 3.1.

**Comment by anurse** [comment]
> Compat Testing:

<chef-kiss.gif> ;). Excellent, thanks!

**Review by davidfowl — APPROVED** [review]

File: src/Components/Server/src/BlazorPack/BlazorPackHubProtocol.cs
Comment (halter73) [inline src/Components/Server/src/BlazorPack/BlazorPackHubProtocol.cs]: @rynowak @pranavkm @mkArtakMSFT FYI I just merged this PR into release/3.1 which changes BlazorPackHubProtocol.

Comment (anurse) [reply src/Components/Server/src/BlazorPack/BlazorPackHubProtocol.cs]: Oof, probably should've gotten sign off from @pranavkm before the merge. Next time :). Can you take a quick look @pranavkm and see if this seems problematic?

**Pull request merged**
